submariner-operator/submariner-metrics-redesign.md

```diff
@@ -0,0 +1,146 @@
+causing a conflict and metrics becoming inaccessible. With the move towards minimizing changes needed in cloud infra for
+submariner, there is a use case to export metrics in a way that doesn't require any of these ports to be opened.
+
+## Proposal
+
+This enhancement proposes a redesign of existing metrics for Submariner Gateway and Globalnet by adding a new metrics proxy
+service. Pods running on the same node can access ports without requiring pods to be opened in the cloud infra. Any queries
+for these metrics will land on the proxy which will always run on the same node.
+
+Note that this proposal doesn't address the issue where metrics port is unavailable on the node. That should be addressed as
+a separate enhancement.
+
+## Design Details
+
+* A new `status.submariner.io/gateway=active` label will be added to gateway node that is currently active.
+* Submarine RouteAgent will set the label on the relevant Node.
+* A new DaemonSet `submariner-metrics` will be created. It will use the NodeSelector `status.submariner.io/gateway=active`
+to make sure it always runs on the active gateway node.
+* This DaemonSet will use `schedulerName: default-scheduler` to make sure that metrics proxy pod moves with the active gateway
+label.
+* DaemonSet will also use toleration to make sure it can always be scheduled on gateway node, irrespective of any taints.
+* `submariner-metrics` Pod will use `netcat` to act as proxy and forward all incoming requests on metrics port to the
+metrics port on the node IP.
+* The proxy pod will run two containers if globlanet is enabled, one each for gateway and globalnet.
+* Existing metrics service for gateway and globalnet will be modified to use the new metrics proxy pod. This will be
+done by using same label selector as metrics DaemonSet.
+
+### Submariner Metrics DaemonSet
+
+```yaml
+---
+apiVersion: apps/v1
+kind: DaemonSet
+metadata:
+  name: submariner-metrics
+spec:
+  selector:
+    matchLabels:
+      app: submariner-metrics
+  template:
+    metadata:
+      labels:
+        app: submariner-metrics
+    spec:
+      nodeSelector:
+        "status.submariner.io/gateway": "active"
+      containers:
+        - name: gateway-metrics-proxy
+          image: quay.io/submariner/nettest:devel
+          imagePullPolicy: IfNotPresent
+          env:
+          - name: NODE_IP
+            valueFrom:
+              fieldRef:
+                fieldPath: status.hostIP
+          - name: SUBM_GW_METRICS_PORT
+            value: "8080"
+          command: ["/bin/sh", "-c", "nc -v -lk -p 8080 -e nc $(NODE_IP) $(SUBM_GW_METRICS_PORT)"]
+        - name: globalnet-metrics-proxy
+          image: quay.io/submariner/nettest:devel
+          imagePullPolicy: IfNotPresent
+          env:
+          - name: NODE_IP
+            valueFrom:
+              fieldRef:
+                fieldPath: status.hostIP
+          - name: SUBM_GN_METRICS_PORT
+            value: "8081"
```

Comment:
Lets use some port from the ephemeral port range.

Reply:
I agree with @sridhargaddam.

Although at this level we don't risk collision (as we're using the pod subnet), the proxy itself can use the same port(s) as the pod it's proxying for simplicity.

On the service we'll still be using the 8080/8081 ports.